File: day-6/day6.js
require( 'fs' ).readFile( 'input.txt', 'utf8', ( err, input ) => {
	if ( err ) {
		console.error( err );
		return;
	}
	console.log( `Start of packet:  ${ find_marker( input, 4 ) }` );
	console.log( `Start of message: ${ find_marker( input, 14 ) }` );
} );

function find_marker( buffer, marker_size ) {
	for ( var pos = marker_size; pos < buffer.length; pos++ ) {
		let chars = {};
		for ( var c = pos - marker_size; c < pos; c++ ) {
			chars[buffer.charAt(c)] = 1;
		}
		if ( Object.keys( chars ).length == marker_size ) return pos;
	}
	return 0;
}

File: day-6/day6.rs
use std::fs;
use std::collections::HashMap;

fn main() {
    let input = fs::read_to_string("input.txt").unwrap();
    println!("Start of packet:  {}", find_marker(&input, 4).unwrap());
    println!("Start of message: {}", find_marker(&input, 14).unwrap());
}

fn find_marker(buffer: &str, marker_size: usize) -> Option<usize> {
    for pos in marker_size..=buffer.len() {
        let mut chars: HashMap<char,usize> = HashMap::new();
        for c in buffer[pos-marker_size..pos].chars() {
            chars.insert(c, 0);
        }
        if chars.len() == marker_size {
            return Some(pos);
        }
    }
    None
}
